enterprise/reporting/error_reporting.go

```diff
@@ -0,0 +1,593 @@
+			return
+		}
+		requestChan := make(chan struct{}, maxConcurrentRequests)
+		loopStart := time.Now()
+		currentMs := time.Now().UTC().Unix() / 60
+
+		getReportsStart := time.Now()
+		reports, reportedAt := edRep.getReports(currentMs, clientName)
+		getReportsTimer.Since(getReportsStart)
+		getReportsCount.Observe(float64(len(reports)))
+		if len(reports) == 0 {
+			select {
+			case <-ctx.Done():
+				edRep.log.Infof("stopping mainLoop for client %s : %s", clientName, ctx.Err())
+				return
+			case <-time.After(edRep.sleepInterval):
+			}
+			continue
+		}
+
+		aggregationStart := time.Now()
+		metrics := edRep.aggregate(reports)
+		aggregateTimer.Since(aggregationStart)
+		getAggregatedReportsCount.Observe(float64(len(metrics)))
+
+		errGroup, errCtx := errgroup.WithContext(ctx)
+		for _, metric := range metrics {
+			metricToSend := metric
+			requestChan <- struct{}{}
+			if errCtx.Err() != nil {
+				// if any of errGroup's goroutines fail - don't send anymore requests for this batch
+				break
+			}
+			errGroup.Go(func() error {
+				err := edRep.sendMetric(errCtx, clientName, metricToSend)
+				if err != nil {
+					edRep.log.Error("Error while sending to Reporting service:", err)
+				}
+				<-requestChan
+				return err
+			})
+		}
+
+		err := errGroup.Wait()
+		if err == nil {
+			sqlStatement := fmt.Sprintf(`DELETE FROM %s WHERE reported_at = %d`, ErrorDetailReportsTable, reportedAt)
+			dbHandle, err := edRep.getDBHandle(clientName)
+			if err != nil {
+				panic(err)
+			}
+			deleteReportsStart := time.Now()
+			_, err = dbHandle.Exec(sqlStatement)
+			errorDetailReportsDeleteQueryTimer.Since(deleteReportsStart)
+			if err != nil {
+				edRep.log.Errorf(`[ Error Detail Reporting ]: Error deleting local reports from %s: %v`, ErrorDetailReportsTable, err)
+			}
+		}
+
+		mainLoopTimer.Since(loopStart)
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(edRep.mainLoopSleepInterval):
+		}
+	}
+}
+
+func (edRep *ErrorDetailReporter) getReports(currentMs int64, clientName string) ([]*types.EDReportsDB, int64) {
+	sqlStatement := fmt.Sprintf(`SELECT reported_at FROM %s WHERE reported_at < %d ORDER BY reported_at ASC LIMIT 1`, ErrorDetailReportsTable, currentMs)
+	var queryMin sql.NullInt64
+	dbHandle, err := edRep.getDBHandle(clientName)
+	if err != nil {
+		panic(err)
+	}
+
+	queryStart := time.Now()
+	err = dbHandle.QueryRow(sqlStatement).Scan(&queryMin)
+	if err != nil && err != sql.ErrNoRows {
+		panic(err)
+	}
+	edRep.minReportedAtQueryTime.Since(queryStart)
+	if !queryMin.Valid {
+		return nil, 0
+	}
+	edSelColumns := strings.Join(ErrorDetailReportsColumns, ", ")
+	sqlStatement = fmt.Sprintf(`SELECT %s FROM %s WHERE reported_at = %d`, edSelColumns, ErrorDetailReportsTable, queryMin.Int64)
+	edRep.log.Debugf("[EdRep] sql statement: %s\n", sqlStatement)
+	var rows *sql.Rows
+	queryStart = time.Now()
+	rows, err = dbHandle.Query(sqlStatement)
+	if err != nil {
+		panic(err)
```

Comment:
When possible we should use prepared statements for queries.

```suggestion
	var queryMin sql.NullInt64
	dbHandle, err := edRep.getDBHandle(clientName)
	if err != nil {
		panic(err)
	}

	queryStart := time.Now()
	err = dbHandle.QueryRowContext(`SELECT reported_at FROM `+ ErrorDetailReportsTable ` WHERE reported_at < $1 ORDER BY reported_at ASC LIMIT 1`, currentMs).Scan(&queryMin)
	if err != nil && err != sql.ErrNoRows {
		panic(err)
	}
	edRep.minReportedAtQueryTime.Since(queryStart)
	if !queryMin.Valid {
		return nil, 0
	}
	edSelColumns := strings.Join(ErrorDetailReportsColumns, ", ")
	sqlStatement = fmt.Sprintf()
	edRep.log.Debugf("[EdRep] sql statement: %s\n", sqlStatement)
	var rows *sql.Rows
	queryStart = time.Now()
	rows, err = dbHandle.Query(
	     `SELECT ` +  strings.Join(ErrorDetailReportsColumns, ", ") + 
	     ` FROM ` + ErrorDetailReportsTable +
	      `WHERE reported_at = $1`, queryMin.Int64)
	if err != nil {
```

Reply:
Done